CapacityLimiter can over-grant tokens (`borrowed_tokens > total_tokens`, `available_tokens` goes negative)

### Things to check first

- [x] I have searched the existing issues and didn't find my bug already reported there

- [x] I have checked that my bug is still present in the latest release


### AnyIO version

4.13.0

### Python version

3.12

### What happened?

On the asyncio backend, a non-blocking acquire can succeed *while the just-freed token is already reserved for a woken waiter*, leaving the limiter with more borrowers than it has tokens. The state becomes:

```
borrowed_tokens  = 2
total_tokens     = 1
available_tokens = -1      # negative
borrowers        = ('B', 'X')
```

Two tasks then hold a single-token limiter concurrently — the exact condition the limiter exists to prevent — and `available_tokens` is negative.

This violates the documented contract of `acquire_nowait` / `acquire_on_behalf_of_nowait`:

> :raises ~anyio.WouldBlock: if there are no tokens available for borrowing

In the window below there is no token available (the freed one is earmarked for the woken waiter), yet the nowait acquire succeeds instead of raising `WouldBlock`.

### How can we reproduce the bug?

Just anyio, no test framework: `python repro.py` prints the corrupt state and raises `AssertionError`:

```python
import anyio
from anyio import CapacityLimiter, WouldBlock

async def main():
    lim = CapacityLimiter(1)                       # capacity: exactly ONE token
    lim.acquire_on_behalf_of_nowait("A")           # A borrows the only token

    async with anyio.create_task_group() as tg:
        async def b():
            await lim.acquire_on_behalf_of("B")    # B waits for the token
        tg.start_soon(b)

        while lim.statistics().tasks_waiting != 1:  # wait until B is parked
            await anyio.sleep(0)

        lim.release_on_behalf_of("A")              # frees the token, notifies B
        # B has been notified but has NOT resumed yet, so it is not in
        # _borrowers and the wait queue is now empty. A non-blocking acquire
        # in this window slips through:
        try:
            lim.acquire_on_behalf_of_nowait("X")  # expected: WouldBlock
            print("X acquired without waiting (should have raised WouldBlock)")
        except WouldBlock:
            print("X correctly blocked")
        # leaving the task group lets B resume and add itself to _borrowers

    print(f"borrowed_tokens  = {lim.borrowed_tokens}  (total_tokens = {lim.total_tokens})")
    print(f"available_tokens = {lim.available_tokens}")
    print(f"borrowers        = {lim.statistics().borrowers}")
    assert lim.borrowed_tokens <= lim.total_tokens, "borrowed_tokens > total_tokens"

anyio.run(main)
```

Output:

```
X acquired without waiting (should have raised WouldBlock)
borrowed_tokens  = 2  (total_tokens = 1)
available_tokens = -1
borrowers        = ('B', 'X')
AssertionError: borrowed_tokens > total_tokens
```
### Cause

In `anyio/_backends/_asyncio.py`:

- `release_on_behalf_of` removes the borrower then calls `_notify_next_waiter()`, which **pops** the next waiter out of `_wait_queue` and does `event.set()`.
- The woken waiter, in `acquire_on_behalf_of`, only adds itself to `_borrowers` **after** `await event.wait()` returns:

  ```python
  await event.wait()
  ...
  self._borrowers.add(borrower)   # runs only once the waiter resumes
  ```

- Between the `event.set()` and the waiter resuming, `_wait_queue` is empty and the waiter is not yet in `_borrowers`. So the nowait guard

  ```python
  if self._wait_queue or len(self._borrowers) >= self._total_tokens:
      raise WouldBlock
  ```

  evaluates to `False` (empty queue, `0 < 1`) and admits a second borrower for the one token.

### Why `Semaphore` is not affected

The sibling `Semaphore` is immune under the identical scenario, a nowait acquire issued in the same window correctly raises `WouldBlock`, because its `release()` hands the freed slot to the woken waiter rather than leaving it generally available. `CapacityLimiter` is the one primitive whose freed-but-reserved token is briefly visible as "available" to a synchronous acquirer.

A fix would close that window, e.g. by reserving the token for the woken waiter at notify time (add the borrower to `_borrowers` when `_notify_next_waiter` sets its event, rather than after it resumes), mirroring the Semaphore/Lock hand-off. The nowait guard would then correctly see no free token.

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -8,11 +8,6 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Changed ``ByteReceiveStream.receive()`` implementations to raise a ``ValueError`` when
   ``max_bytes`` is not a positive integer
   (`#1191 <https://github.com/agronholm/anyio/pull/1191>`_)
-- Fixed ``CapacityLimiter.total_tokens`` rejecting ``float("inf")`` when the limiter was
-  instantiated outside of an event loop. The adapter setter checked for infinity by
-  identity (``value is math.inf``), so only the exact ``math.inf`` singleton was accepted,
-  while every backend setter (using ``math.isinf()``) accepts any positive infinity
-  (`#1189 <https://github.com/agronholm/anyio/pull/1189>`_; PR by @greymoth-jp).
 - Fixed ``to_process.run_sync()`` deadlocking when the worker function writes enough data
   to ``sys.stderr`` to fill the (undrained) pipe buffer. The worker process now redirects
   ``sys.stderr`` to ``os.devnull`` as well, matching the documented behavior
@@ -23,6 +18,11 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Fixed ``anyio.open_process()`` (and ``run_process()``) ignoring the ``extra_groups``
   argument, as it mistakenly passed the value of the ``group`` argument instead
   (`#1209 <https://github.com/agronholm/anyio/pull/1209>`_)
+- Fixed ``CapacityLimiter.total_tokens`` rejecting ``float("inf")`` when the limiter was
+  instantiated outside of an event loop. The adapter setter checked for infinity by
+  identity (``value is math.inf``), so only the exact ``math.inf`` singleton was accepted,
+  while every backend setter (using ``math.isinf()``) accepts any positive infinity
+  (`#1189 <https://github.com/agronholm/anyio/pull/1189>`_; PR by @greymoth-jp).
 
 **4.14.1**
 

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -50,6 +50,13 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
   * ``anyio.Lock``
   * ``anyio.ResourceGuard``
   * ``anyio.Semaphore``
+- Fixed ``CapacityLimiter`` on the asyncio backend over-granting tokens
+  (``borrowed_tokens`` exceeding ``total_tokens`` and ``available_tokens`` going
+  negative) when a non-blocking acquire was made in the window between a token
+  being released and the notified waiter resuming. The freed token is now
+  reserved for the woken waiter right away, so the non-blocking acquire correctly
+  raises ``WouldBlock``
+  (`#1170 <https://github.com/agronholm/anyio/issues/1170>`_; PR by @gaoflow)
 - Fixed cancellation exception escaping a cancel scope when triggered via
   ``check_cancelled()`` in a worker thread
   (`#1113 <https://github.com/agronholm/anyio/issues/1113>`_)

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -8,6 +8,11 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Changed ``ByteReceiveStream.receive()`` implementations to raise a ``ValueError`` when
   ``max_bytes`` is not a positive integer
   (`#1191 <https://github.com/agronholm/anyio/pull/1191>`_)
+- Fixed ``CapacityLimiter.total_tokens`` rejecting ``float("inf")`` when the limiter was
+  instantiated outside of an event loop. The adapter setter checked for infinity by
+  identity (``value is math.inf``), so only the exact ``math.inf`` singleton was accepted,
+  while every backend setter (using ``math.isinf()``) accepts any positive infinity
+  (`#1189 <https://github.com/agronholm/anyio/pull/1189>`_; PR by @greymoth-jp).
 - Fixed ``to_process.run_sync()`` deadlocking when the worker function writes enough data
   to ``sys.stderr`` to fill the (undrained) pipe buffer. The worker process now redirects
   ``sys.stderr`` to ``os.devnull`` as well, matching the documented behavior
@@ -18,11 +23,13 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Fixed ``anyio.open_process()`` (and ``run_process()``) ignoring the ``extra_groups``
   argument, as it mistakenly passed the value of the ``group`` argument instead
   (`#1209 <https://github.com/agronholm/anyio/pull/1209>`_)
-- Fixed ``CapacityLimiter.total_tokens`` rejecting ``float("inf")`` when the limiter was
-  instantiated outside of an event loop. The adapter setter checked for infinity by
-  identity (``value is math.inf``), so only the exact ``math.inf`` singleton was accepted,
-  while every backend setter (using ``math.isinf()``) accepts any positive infinity
-  (`#1189 <https://github.com/agronholm/anyio/pull/1189>`_; PR by @greymoth-jp).
+- Fixed ``CapacityLimiter`` on the asyncio backend over-granting tokens
+  (``borrowed_tokens`` exceeding ``total_tokens`` and ``available_tokens`` going
+  negative) when a non-blocking acquire was made in the window between a token
+  being released and the notified waiter resuming. The freed token is now
+  reserved for the woken waiter right away, so the non-blocking acquire correctly
+  raises ``WouldBlock``
+  (`#1170 <https://github.com/agronholm/anyio/issues/1170>`_; PR by @gaoflow)
 
 **4.14.1**
 
@@ -94,13 +101,6 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
   * ``anyio.Lock``
   * ``anyio.ResourceGuard``
   * ``anyio.Semaphore``
-- Fixed ``CapacityLimiter`` on the asyncio backend over-granting tokens
-  (``borrowed_tokens`` exceeding ``total_tokens`` and ``available_tokens`` going
-  negative) when a non-blocking acquire was made in the window between a token
-  being released and the notified waiter resuming. The freed token is now
-  reserved for the woken waiter right away, so the non-blocking acquire correctly
-  raises ``WouldBlock``
-  (`#1170 <https://github.com/agronholm/anyio/issues/1170>`_; PR by @gaoflow)
 - Fixed cancellation exception escaping a cancel scope when triggered via
   ``check_cancelled()`` in a worker thread
   (`#1113 <https://github.com/agronholm/anyio/issues/1113>`_)

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -2091,13 +2091,7 @@ class CapacityLimiter(BaseCapacityLimiter):
         return self._total_tokens - len(self._borrowers)
 
     def _notify_next_waiter(self) -> None:
-        """Hand a free token to the next task in line, if any.
-
-        The token is reserved for the woken task right away (the borrower is
-        added to ``_borrowers`` before its event is set) so that the freed token
-        is never momentarily visible as available to a non-blocking acquire made
-        before the woken task has resumed.
-        """
+        """Hand a free token to the next task in line, if any."""
         if self._wait_queue and len(self._borrowers) < self._total_tokens:
             borrower, event = self._wait_queue.popitem(last=False)
             self._borrowers.add(borrower)
@@ -2132,16 +2126,10 @@ class CapacityLimiter(BaseCapacityLimiter):
             except BaseException:
                 self._wait_queue.pop(borrower, None)
                 if event.is_set():
-                    # The token was already reserved for us when we were
-                    # notified; since we're bailing out, release it and pass it
-                    # on to the next waiter.
                     self._borrowers.discard(borrower)
                     self._notify_next_waiter()
 
                 raise
-
-            # The token was reserved for us by the notifying side, so there's no
-            # need to add ourselves to the borrowers here.
         else:
             try:
                 await AsyncIOBackend.cancel_shielded_checkpoint()

--- a/src/anyio/_backends/_asyncio.py
+++ b/src/anyio/_backends/_asyncio.py
@@ -2042,7 +2042,8 @@ class CapacityLimiter(BaseCapacityLimiter):
 
         # Notify waiting tasks that they have acquired the limiter
         while self._wait_queue and waiters_to_notify:
-            event = self._wait_queue.popitem(last=False)[1]
+            borrower, event = self._wait_queue.popitem(last=False)
+            self._borrowers.add(borrower)
             event.set()
             waiters_to_notify -= 1
 
@@ -2055,9 +2056,16 @@ class CapacityLimiter(BaseCapacityLimiter):
         return self._total_tokens - len(self._borrowers)
 
     def _notify_next_waiter(self) -> None:
-        """Notify the next task in line if this limiter has free capacity now."""
+        """Hand a free token to the next task in line, if any.
+
+        The token is reserved for the woken task right away (the borrower is
+        added to ``_borrowers`` before its event is set) so that the freed token
+        is never momentarily visible as available to a non-blocking acquire made
+        before the woken task has resumed.
+        """
         if self._wait_queue and len(self._borrowers) < self._total_tokens:
-            event = self._wait_queue.popitem(last=False)[1]
+            borrower, event = self._wait_queue.popitem(last=False)
+            self._borrowers.add(borrower)
             event.set()
 
     def acquire_nowait(self) -> None:
@@ -2089,11 +2097,16 @@ class CapacityLimiter(BaseCapacityLimiter):
             except BaseException:
                 self._wait_queue.pop(borrower, None)
                 if event.is_set():
+                    # The token was already reserved for us when we were
+                    # notified; since we're bailing out, release it and pass it
+                    # on to the next waiter.
+                    self._borrowers.discard(borrower)
                     self._notify_next_waiter()
 
                 raise
 
-            self._borrowers.add(borrower)
+            # The token was reserved for us by the notifying side, so there's no
+            # need to add ourselves to the borrowers here.
         else:
             try:
                 await AsyncIOBackend.cancel_shielded_checkpoint()

--- a/tests/test_synchronization.py
+++ b/tests/test_synchronization.py
@@ -994,13 +994,12 @@ class TestCapacityLimiter:
         with pytest.raises(WouldBlock):
             limiter.acquire_on_behalf_of_nowait(borrower2)
 
-    @pytest.mark.parametrize("anyio_backend", asyncio_params)
     async def test_nowait_acquire_after_release_does_not_oversubscribe(self) -> None:
         # Regression test for #1170: a non-blocking acquire issued in the window
         # between releasing a token (which notifies the next waiter) and that
         # waiter actually resuming must not slip through, as the freed token is
-        # already reserved for the woken waiter. This window is specific to the
-        # asyncio backend (the trio backend wraps trio's own CapacityLimiter).
+        # already reserved for the woken waiter. The over-granting bug was
+        # specific to the asyncio backend, but the invariant holds on both.
         limiter = CapacityLimiter(1)
         limiter.acquire_on_behalf_of_nowait("A")
 

--- a/tests/test_synchronization.py
+++ b/tests/test_synchronization.py
@@ -929,3 +929,32 @@ class TestCapacityLimiter:
             tg.cancel_scope.cancel()
 
         pytest.fail("The second borrower failed to acquire the limiter")
+
+    @pytest.mark.parametrize("anyio_backend", asyncio_params)
+    async def test_nowait_acquire_after_release_does_not_oversubscribe(self) -> None:
+        # Regression test for #1170: a non-blocking acquire issued in the window
+        # between releasing a token (which notifies the next waiter) and that
+        # waiter actually resuming must not slip through, as the freed token is
+        # already reserved for the woken waiter. This window is specific to the
+        # asyncio backend (the trio backend wraps trio's own CapacityLimiter).
+        limiter = CapacityLimiter(1)
+        limiter.acquire_on_behalf_of_nowait("A")
+
+        async def waiter() -> None:
+            await limiter.acquire_on_behalf_of("B")
+
+        async with create_task_group() as tg:
+            tg.start_soon(waiter)
+            await wait_all_tasks_blocked()
+            assert limiter.statistics().tasks_waiting == 1
+
+            # Frees the only token and reserves it for the still-parked "B"; no
+            # token is available, so the nowait acquire must raise WouldBlock.
+            limiter.release_on_behalf_of("A")
+            with pytest.raises(WouldBlock):
+                limiter.acquire_on_behalf_of_nowait("X")
+
+        assert limiter.borrowed_tokens <= limiter.total_tokens
+        assert limiter.available_tokens >= 0
+        assert limiter.statistics().borrowers == ("B",)
+        limiter.release_on_behalf_of("B")

